apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,341 @@
+their performance. To understand how these transforms can be compiled, it's
+essential to discuss the workings of the three layers within the compilation
+stack, which are namely `dynamo`, `aot_autograd`, and `inductor`.
+
+The `dynamo` and `aot_autograd` layers primarily focus on capturing the
+computation graph and converting the captured operations into more basic operations.
+This captured graph is then passed to `inductor`, the compiler. `inductor` then applies
+various optimizations passes before generating specialized code.
+
+To gain insight into the different stages of this stack,
+let's compile a simple program in debug mode.
+
+```python
+# Run this file with `TORCH_COMPILE_DEBUG=1`
+
+import torch
+
+def fn(x):
+    return torch.sin(x) + torch.square(x)
+
+torch.compile(fn)(torch.randn(3, 3))
+```
+
+**dynamo**: The primary responsibility of dynamo is to trace the Python program
+and convert it into the FX graph format. The FX graph generated by `dynamo`
+represents PyTorch operations from the public API, such as `torch.sin`.
+Below, you can observe the graph captured by `dynamo`.
+
+```python
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_ : torch.Tensor):
+        l_x_ = L_x_
+
+        # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+        sin = torch.sin(l_x_)
+        square = torch.square(l_x_);  l_x_ = None
+        add = sin + square;  sin = square = None
+        return (add,)
+```
+
+**aot_autograd**: `aot_autograd` retraces all PyTorch operations to
+produce a lower-level FX graph using `aten` functions (from the private API).
+Additionally, `aot_autograd` decomposes composite operations into primitive
+operations. For instance, a composite operation like `torch.square` is traced
+down to `torch.pow(x, 2)`.
+
+Moreover, `aot_autograd` also manages the creation of the backward graph when
+requested. This is useful for transforms like `grad`, `vjp`, etc. Below, you can see the
+graph generated by `aot_autograd` for the above program.
+
+```python
+def forward(self, arg0_1: f32[3, 3]):
+    # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+    sin: f32[3, 3] = torch.ops.aten.sin.default(arg0_1)
+    pow_1: f32[3, 3] = torch.ops.aten.pow.Tensor_Scalar(arg0_1, 2);  arg0_1 = None
+    add: f32[3, 3] = torch.ops.aten.add.Tensor(sin, pow_1);  sin = pow_1 = None
+    return (add,)
+
+```
+
+**inductor**: As discussed above, it is inductor's job to apply optimisations and
+generate specialised code. In this case, it has fused `sin` and `square` to run
+within the same `for`-loop. This allows the generated program to do more compute
+per read/write effectively improving the memory bandwith utilization.
+
+```c++
+#include "/tmp/torchinductor_kshiteej/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
+extern "C" void kernel(const float* in_ptr0,
+                       float* out_ptr0)
+{
+    {
```

Comment:
```suggestion
                       float* out_ptr0){
```
and dedent the rest of the code.
Consider also removing the static_casts from the code below for readability.